// angular-google-charts/esm5/lib/components/chart-editor/chart-editor-ref.js
import { Subject } from 'rxjs';
var ChartEditorRef = /** @class */ (function () {
    function ChartEditorRef(editor) {
        this.editor = editor;
        this.doneSubject = new Subject();
        this.addEventListeners();
    }
    /**
     * Gets an observable that is notified when the dialog is saved.
     * Emits either the result if the dialog was saved or `null` if editing was cancelled.
     */
    ChartEditorRef.prototype.afterClosed = function () {
        return this.doneSubject.asObservable();
    };
    /**
     * Stops editing the chart and closes the dialog.
     */
    ChartEditorRef.prototype.cancel = function () {
        this.editor.closeDialog();
    };
    ChartEditorRef.prototype.addEventListeners = function () {
        var _this = this;
        google.visualization.events.addOneTimeListener(this.editor, 'ok', function () {
            google.visualization.events.removeAllListeners(_this.editor);
            var updatedChartWrapper = _this.editor.getChartWrapper();
            _this.doneSubject.next(updatedChartWrapper);
            _this.doneSubject.complete();
        });
        google.visualization.events.addOneTimeListener(this.editor, 'cancel', function () {
            google.visualization.events.removeAllListeners(_this.editor);
            _this.doneSubject.next(null);
            _this.doneSubject.complete();
        });
    };
    return ChartEditorRef;
}());
export { ChartEditorRef };
